### PR TITLE
Update controller runtime

### LINKED ISSUE #3300: Bug: Crash when specifying a single namespace for AZURE_TARGET_NAMESPACES
**Version of Azure Service Operator**

2.3.0

**Describe the bug**

Specifying a single namespace in `AZURE_TARGET_NAMESPACES` causes ASO to CrashLoopBackOff with the following log message:
```bash
❯ kubectl logs azureserviceoperator-controller-manager-7f4db8b7c-sh8jp
I0914 15:36:59.861906       1 main.go:29] setup "msg"="Launching with flags" "flags"="MetricsAddr: 0.0.0.0:8080, HealthAddr: :8081, EnableLeaderElection: true, CRDPatterns: , PreUpgradeCheck: false"
E0914 15:36:59.920709       1 setup.go:129] setup "msg"="unable to create manager" "error"="must specify more than one namespace to use multi-namespace cache"
```

**To Reproduce**

1. Specify a single value for `AZURE_TARGET_NAMESPACES` 
2. Watch it crash

**Expected behavior**
Not crash

**Screenshots**
N/A

**Additional context**
values used during install:
```
❯ helm get values aso
USER-SUPPLIED VALUES:
azureClientId: {real value}
azureOperatorMode: watchers
azureSubscriptionId: {real value}
azureTargetNamespaces: testns
azureTenantId: {real value}
createAzureOperatorSecret: true
multitenant:
  enable: true
useWorkloadIdentityAuth: true
```

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

As reported in #3300 and as identified by @matthchr, our upgrade to controller-runtime v0.15.0 in ASO 2.3.0 introduced a bug. Upgrading to the newer release, v0.16.x should remedy the problem.

Closes #3300

**How does this PR make you feel**:
gif

### CODE REVIEW

**Review by super-harsh — APPROVED** [review]

File: v2/internal/testcommon/kube_test_context_envtest.go [hunk 1]
```diff
@@ -102,14 +104,19 @@ func createSharedEnvTest(cfg testConfig, namespaceResources *namespaceResources)
 
 	var cacheFunc cache.NewCacheFunc
 	if cfg.TargetNamespaces != nil && cfg.OperatorMode.IncludesWatchers() {
-		cacheFunc = cache.MultiNamespacedCacheBuilder(cfg.TargetNamespaces)
+		cacheFunc = func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
```
Comment (super-harsh) [inline v2/internal/testcommon/kube_test_context_envtest.go:107]: Consider extracting this out in a method and call it here directly?

File: v2/cmd/controller/app/setup.go [hunk 2]
```diff
@@ -111,18 +113,29 @@ func SetupControllerManager(ctx context.Context, setupLog logr.Logger, flgs Flag
 
 	var cacheFunc cache.NewCacheFunc
 	if cfg.TargetNamespaces != nil && cfg.OperatorMode.IncludesWatchers() {
-		cacheFunc = cache.MultiNamespacedCacheBuilder(cfg.TargetNamespaces)
+		cacheFunc = func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
```
Comment (super-harsh) [inline v2/cmd/controller/app/setup.go:116]: Same: Consider extracting this out in a method and call it here directly?

File: v2/internal/testcommon/kube_test_context_envtest.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/internal/testcommon/kube_test_context_envtest.go:107]: I did consider that. It's passed as a callback that's invoked later and needs to capture locals from the current context. I'm not sure the refactored method would be any clearer.

Comment (theunrepentantgeek) [reply v2/internal/testcommon/kube_test_context_envtest.go:107]: I'll check out the possibility of a refactor offline; want to get this merged to clear dependabot PRs.

**Pull request merged**